File: grade01.java
import java.util.Scanner;

public class grade01{
    public static void main(String[] args) {
        result();
        //rules:
        //Exam 1 = 20%
        //Exam 2 = 30%
        //Homework, assignments(not separate) = 50%% *
        //labs total are 10(needed user input)
        //exam 1 & 2 is 100 (goto the result function where it says examScore1 and 2)
        //the max score was 100 and the final can't get higher than 100
        //no late attendance gets a 5 point credit reward.
        //more than 50 total late attendance, the 50% maybe reduced to 40% of the total grade is reduced more by 10% after the homework equation
        //conclusion is getting 2 or even one can delete 2-6 points. even if you got a 100
        //rough rule
    }



    public static double AttendenceLab(int labsAttended, int total, double homeworkRule){
        if (labsAttended < (total/2)){
            homeworkRule = homeworkRule * 0.9;//if labs attended are less than half, 10% deduction
        }
        else if (labsAttended == total){
            homeworkRule = homeworkRule + 5;//if all labs attended, 5 extra credit
        }
        return homeworkRule;//we will call them in the homework function
    }
    
    public static double homework(int total, int homeworkScore, int labsAttended){//this one 
        double homeworkRule = 50 * (homeworkScore * 10 + 4 * labsAttended) / (total * 10 + 4 * 10);
        homeworkRule = AttendenceLab(labsAttended, total, homeworkRule);//call the attendence lab function
        return homeworkRule;
    }
    
    public static double exam01(int examScore1){
        return 20 * examScore1 / 100.0;
    }
    
    public static double exam02(int examScore2){
        if (examScore2 > 94){
            examScore2 += 10;
            if (examScore2 > 100){
                examScore2 = 100;
            }
        }
        return 30 * examScore2 / 100.0;
    }
    
    public static double finals(double homework, double exam01, double exam02){
        double finalScore = homework + exam01 + exam02;
        return finalScore;
    }



    public static void result(){//the body of the program
        int examScore1 = 100;
        int examScore2 = 100;
        Scanner input = new Scanner(System.in);
        System.out.print("Total labs and homework: ");
        int total = input.nextInt();
        System.out.print("Enter number of labs attended: ");
        int labsAttended = input.nextInt(); 
        System.out.print("Enter total homework score: ");
        double homeworkScore = input.nextDouble();
        double homework = homework(total, (int)homeworkScore, labsAttended);
        double exam01 = exam01(examScore1);
        double exam02 = exam02(examScore2);
        double finalScore = finals(homework, exam01, exam02);
        System.out.println("Final score: " + finalScore);
        if(finalScore > 100){ //score can't be more than 100(no extra credit)
            System.out.println("Final score: 100");
        } 
        else {
            System.out.println("Final score: " + finalScore);
        }
    input.close();
    }
}

File: grade01.py



def exam1Final(exam1Grade, exam1Weight):
        return exam1Grade * exam1Weight / 100

def exam2Final(exam2Grade, exam2Weight):
        return exam2Grade * exam2Weight / 100

def homeworkFinal(homeworkGrade, homeworkWeight, totalAssignments, lateAssignments):
        return homeworkWeight * (homeworkGrade * totalAssignments + (totalAssignments - lateAssignments) * 4) / (totalAssignments * 100)
'''
def annoyingPeople_rules(homeworkGrade):
    while True:
        if homeworkGrade > 100:
            print("Please don't be a jerk and enter a valid grade than is less than 100 please... ")
            Grade = float(input("Please enter your grade, 100 is the max can't do more: "))
        elif homeworkGrade < 0:
            print("now lesser than 0? really? fine then... ")
            Grade = float(input("Please enter your grade 100 is the max, AND 0 IS THE MIN can't do more!!!: "))
        else:
            break
    return Grade/10 
'''
def addeditionalRules(lateAssignments, totalAssignments, homeworkOutput):
    #no late attendance = + 5
    #50 percent late = -10%
    if lateAssignments == 0:
        return homeworkOutput + 5
    elif lateAssignments >= totalAssignments / 2:
        return homeworkOutput*(10/100)
    return homeworkOutput


def main():
    exam1Weight = float(input("Please enter the weight for exam 1: "))
    exam2Weight = float(input("Please enter the weight for exam 2: "))
    homeworkWeight = float(input("Please enter the weight for homework and assignments: "))
    totalAssignments = int(input("Please enter the total number of assignments/homework and labs: "))

    homeworkGrade = float(input("Please enter your homework grade 100 is the max can't do more: "))
    lateAssignments = int(input("Please enter the number of no late assignments: "))

    exam1Grade = float(input("Please enter your exam 1 grade 100 is the max can't do more: "))
    exam2Grade = float(input("Please enter your exam 2 grade 100 is the max can't do more: "))
    exam1Output = exam1Final(exam1Grade, exam1Weight)
    exam2Output = exam2Final(exam2Grade, exam2Weight)
    #annoyingPeople_rules(homeworkGrade)
    homeworkOutput = homeworkFinal(homeworkGrade, homeworkWeight, totalAssignments, lateAssignments)
    addeditionalRules(lateAssignments, totalAssignments, homeworkOutput)
    finalGrade = exam1Output + exam2Output + homeworkOutput
    if finalGrade > 100:
         finalGrade = 100
    print(f"Your final output is {finalGrade}")
    


main()
#exam 1 = 20%
#exam 2 = 30%
#assignment and homework = 50%
#user input is asked for total labs and assignments.
#if half the assignments are late, deduct 10% from the homework grade.\
#if there are no late assignments, add 5 points to the homework grade.
#I gave out most of it
# too much rules